pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -337,8 +339,83 @@ func (i *InfraProvider) DestroyBootstrap(dir string) error {
 	return nil
 }
 
+type machineManifest struct {
+	Status struct {
+		Addresses []clusterv1.MachineAddress `yaml:"addresses"`
+	} `yaml:"status"`
+}
+
+// extractIPAddress extracts the IP address from a machine manifest file in a
+// provider-agnostic way by reading only the "status" stanza, which should be
+// present in all providers.
+func extractIPAddress(manifestPath string) (string, error) {
+	data, err := os.ReadFile(manifestPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read machine manifest %s: %w", manifestPath, err)
+	}
+	var manifest machineManifest
+	if err := yaml.Unmarshal(data, &manifest); err != nil {
+		return "", fmt.Errorf("failed to unmarshal manifest %s: %w", manifestPath, err)
+	}
+
+	var ipAddr string
+	for _, addr := range manifest.Status.Addresses {
+		switch addr.Type {
+		case clusterv1.MachineExternalIP:
+			ipAddr = addr.Address
```

Comment:
nit: not that there should be many, but technically as long as this is not empty could break here (or put at the top of the loop?).

Reply:
Indeed. Will fix in the next push with other fixes.

Reply:
As we discussed on Slack, a `break` here is redundant since it affects the `switch` not the `for`. So I'll leave this as-is.